proposed/2022/Multiple-Equivalent-Framework-Support-TFM-As-Aliases.md

```diff
@@ -0,0 +1,220 @@
+In some cases, it is convenient to multi target based on different pivots such as target appplication version or maybe the specific runtime, as packages can contain runtime assets.
+This proposals covers the steps that need to be taken to enable that capability.
+
+## Motivation
+
+Allowing the same framework to be using multiple times in restore would allow customers to use the same project to generate runtime specific assemblies based on the same target framework.
+It would also allow scenarios such as different builds for VSIXes targeting different Visual Studio versions.
+
+## Explanation
+
+### Functional explanation
+
+#### TargetFramework values are aliases today
+
+NuGet currently treats `TargetFramework` as aliases. For example, the following is a valid project.
+
+```xml
+<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <TargetFramework>banana</TargetFramework>
+  </PropertyGroup>
+
+  <PropertyGroup Condition=" '$(TargetFramework)' == 'banana' ">
+    <TargetFrameworkIdentifier>.NETFramework</TargetFrameworkIdentifier>
+    <TargetFrameworkVersion>v4.7.2</TargetFrameworkVersion>
+    <TargetFrameworkMoniker>.NETFramework,Version=v4.7.2</TargetFrameworkMoniker>
+  </PropertyGroup>
+
+</Project>
+```
+
+```console
+C:\Code\Temp\Aliases [main]> dotnet restore
+  Determining projects to restore...
+  Restored C:\Code\Temp\Aliases\Aliases.csproj (in 47 ms).
+
+C:\Code\Temp\Aliases [main]> dotnet build --framework banana
+MSBuild version 17.3.0+92e077650 for .NET
+  Determining projects to restore...
+  All projects are up-to-date for restore.
+  Aliases -> C:\Code\Temp\Aliases\bin\Debug\banana\Aliases.dll
+
+Build succeeded.
+    0 Warning(s)
+    0 Error(s)
+
+Time Elapsed 00:00:00.27
+
+C:\Code\Temp\Aliases [main]> dotnet publish --framework banana
+MSBuild version 17.3.0+92e077650 for .NET
+  Determining projects to restore...
+  All projects are up-to-date for restore.
+  Aliases -> C:\Code\Temp\Aliases\bin\Debug\banana\Aliases.dll
+  Aliases -> C:\Code\Temp\Aliases\bin\Debug\banana\publish\
+```
+
+The missing part is allowing the same framework to be targeting among these aliases.
+
+The following scenarios would be enabled:
+
+```xml
+<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <TargetFrameworks>latestnet-linux;latestnet-mac</TargetFrameworks>
+  </PropertyGroup>
+
+  <PropertyGroup Condition=" '$(TargetFramework)' == 'latestnet-mac' OR '$(TargetFramework)' == 'latestnet-linux'">
+    <TargetFrameworkIdentifier>.NETCoreApp</TargetFrameworkIdentifier>
+    <TargetFrameworkVersion>v6.0</TargetFrameworkVersion>
+    <TargetFrameworkMoniker>.NETCoreApp,Version=v6.0</TargetFrameworkMoniker>
+  </PropertyGroup>
+  
+  </Project>
+```
+
+Alternatively, the VS extension scenario would be something like:
+
+```xml
+<Project Sdk="Microsoft.VisualStudio.Extensibility.Sdk">
+
+  <PropertyGroup>
+    <TargetFrameworks>vs17.2;vs17.3</TargetFrameworks>

```

Comment:
I think my doc needs work as it probably reads as I'm suggesting the values here. It is really up to the project author/SDK author to supply values that imply the correct thing from what they are attempting to do. 

The idea is that NuGet and other .NET build component do not need to interpret that value

Reply:
Hypothetical scenario. I'm talking as an example usage. Not to be taken literally. High 10-foot point of view of the feature. In Design terms.

Reply:
While [msbuild static graph](https://github.com/dotnet/msbuild/blob/main/documentation/specs/static-graph.md) introduces MSBuild properties `InnerBuildProperty` and `InnerBuildPropertyValues`, so in theory the VSIX SDK could use `<TargetVS>vs17.8;vs16.11</TargetVS>`, the [msbuild reserved and well known properties docs](https://learn.microsoft.com/en-us/visualstudio/msbuild/msbuild-reserved-and-well-known-properties?view=vs-2022) explicitly states  that this is only applicable for graph builds. Therefore, this doesn't appear to be usable general feature (for example, `dotnet build` without the `-graphBuild` argument).

Additionally, many dotnet CLI commands already have a `--framework` argument on many commands, like `dotnet build -f net8.0`, `dotnet publish -f net8.0`, `dotnet test -f net8.0`.  Even if `<TargetFramework>` isn't a perfect fit for things like platforms/runtimes/VS versions, by using `TargetFrameworks`, all these commands will probably work without modification. On the other hand, if we use some new MSBuild property instead, then there will be a bunch of questions about what CLI arguments should be used when one inner build is desired, what to name the argument, etc.  Then it will be confusing why ProjectA needs a different CLI argument to build/test/whatever, compared to ProjectB.

So, re-using `TargetFrameworks` might not be ideal, but it simplifies several things, so from a practical point of view, I feel like it's better than any alternative.